pkg/innerring/subnet.go

```diff
@@ -0,0 +1,332 @@
+package innerring
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"errors"
+	"fmt"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/mempoolevent"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	irsubnet "github.com/nspcc-dev/neofs-node/pkg/innerring/processors/subnet"
+	morphsubnet "github.com/nspcc-dev/neofs-node/pkg/morph/client/subnet"
+	"github.com/nspcc-dev/neofs-node/pkg/morph/event"
+	subnetevents "github.com/nspcc-dev/neofs-node/pkg/morph/event/subnet"
+	"github.com/nspcc-dev/neofs-node/pkg/util"
+	"github.com/nspcc-dev/neofs-sdk-go/owner"
+	"github.com/nspcc-dev/neofs-sdk-go/subnet"
+	subnetid "github.com/nspcc-dev/neofs-sdk-go/subnet/id"
+	"github.com/panjf2000/ants/v2"
+	"go.uber.org/zap"
+)
+
+// IR server's component to handle Subnet contract notifications.
+type subnetHandler struct {
```

Comment:
It looks like `subnetHandler` is literally subnet processor: a structure that contains worker pool, external dependencies such as client or validators. Is there a reason to define it such way?

Reply:
I've decided to have all components related to event listening in a single layer of the app. It doesn't make sense to pass contract address, worker pool and so on to processor and get parser/handlers back. If IR server controls event listener, then it should do all this work on its side.

Reply:
I think it does make sense to group all operations related to the contract in separate entites, so the server is mostly application constructor and storage of the application state. All event handlings are done in processor, so it is quite easy to find, lookup and debug what application does when event appears. E.g. you can simply find out all actions IR is doing after new epoch event in corresponding processor:
https://github.com/nspcc-dev/neofs-node/blob/011d0f605b8149d3dc42b660f214b7f39fb4a116/pkg/innerring/processors/netmap/process_epoch.go#L55-L60

Anyway, like it or not, all other events processed by separate processor entities, so it is kinda architecture decision. It will be more consistent, if subnet events will be processed the same way. Let's rework that in https://github.com/nspcc-dev/neofs-node/issues/991